### PR TITLE
Add sfx receiver

### PR DESCRIPTION
**Pull request by pjanotti**
Add a receiver supporting SignalFx metrics.

### CODE REVIEW

File: receiver/signalfxreceiver/receiver.go [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+	}
+	if httpStatusCode == http.StatusInternalServerError {
+		traceStatus.Code = trace.StatusCodeInternal
+	}
+	if err != nil {
+		traceStatus.Message = err.Error()
+	}
+	reqSpan.SetStatus(traceStatus)
+
+	r.logger.Warn(
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/receiver.go]: Do we need this warning? This is business as usual. If someone sends us a lot of bad requests we will flood the logs. I think it will be better to keep internal metric/counter of bad requests instead and make this "Debug" level.

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+		labelKeys, labelValues := buildLabelKeysAndValues(sfxDataPoint.Dimensions)
+		descriptor := buildDescriptor(sfxDataPoint, labelKeys, metricType)
+		ts := &metricspb.TimeSeries{
+			// TODO: StartTimestamp can be set if each cumulative time series are
+			//  	tracked but right now it is not clear if it brings benefits.
+			//		Perhaps as an option so cost is "pay for play".
+			LabelValues: labelValues,
+			Points:      []*metricspb.Point{point},
+		}
+		metric := &metricspb.Metric{
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: How about `sfxDataPoint.Properties` field? Do we ignore this? I think this corresponds to `Resource` attributes (I may be wrong).

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+func convertType(
+	sfxDataPoint *sfxpb.DataPoint,
+) (descType metricspb.MetricDescriptor_Type, err error) {
+
+	// Combine metric type with the actual data point type
+	sfxMetricType := sfxDataPoint.GetMetricType()
+	sfxDatum := sfxDataPoint.Value
+	if sfxDatum == nil {
+		return descType, errSFxNilDatum
```
Comment (tigrannajaryan) [inline receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: What's the value of `descType` at this point? Is it the zero value? Perhaps better to specify it explicitly to make it clear?

**Comment by tigrannajaryan** [comment]
Please rebase and resolve the conflicts.

File: receiver/signalfxreceiver/receiver.go [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+	if config.Endpoint == "" {
+		return nil, errEmptyEndpoint
+	}
+
+	// Handle config zero values.
+	if config.Name() == "" {
+		config.SetType(typeStr)
+		config.SetName(typeStr)
+	}
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: Config type is has public members, in principle they may come with respective values either via factory or the New function. In practice it shouldn't be that relevant but since the name is used for metrics and logging I want to be sure that there is some value.

File: receiver/signalfxreceiver/receiver.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+	mux := mux.NewRouter()
+	mux.HandleFunc("/v2/datapoint", r.handleReq)
+	r.server.Handler = mux
+
+	return r, nil
+}
+
+// StartMetricsReception tells the receiver to start its processing.
+// By convention the consumer of the data received is set at creation time.
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: This is an old conversation regarding where/when to set the next consumer: one option was to set it at start or in a separate method so it would become part of the interface, the other was to do it at creation time. In the end of the discussion people preferred to set it at creation time, this sentence tries to capture that. BTW, it is the same in the interface definition on core repo, perhaps worth an update there to.

File: receiver/signalfxreceiver/receiver.go (hunk 1 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: I will move it to debug, regarding the metrics I am on the fence if I add a custom one now or if we do the work of having them on observability so they become standard/available to others.

File: receiver/signalfxreceiver/receiver.go [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+		return
+	}
+
+	if req.Header.Get("Content-Type") != "application/x-protobuf" {
+		r.failRequest(resp, http.StatusUnsupportedMediaType, responseInvalidContentType, nil, span)
+		return
+	}
+
+	encoding := req.Header.Get("Content-Encoding")
+	if encoding != "" && encoding != "gzip" {
```
[the remark this answers is not in the document]
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: sure

File: receiver/signalfxreceiver/receiver.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply receiver/signalfxreceiver/receiver.go]: Config loading code always calls SetName after creating the config: https://github.com/open-telemetry/opentelemetry-collector/blob/4e99ac7468fe44c4456acacad28e2fda31beb7ae/config/config.go#L359

Factory's CreateDefaultConfig also set's the name.

This should not be an issue in any of the code paths that exist. It may happen only if `Config` is created manually (which AFAIK we only do in tests) and someone forgets to set the field, but then there is other things that one can forget. I would say if you are going to create `Config` manually then make sure you set the right fields.

To me this code looks suspicious because it makes me think we have a bug elsewhere where it is supposed to be setting the right name when the Config is created. :-)

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go (hunk 2 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: The type that I'm using is from proto and it doesn't have `Properties`:
```golang
type DataPoint struct {
	Source           *string      `protobuf:"bytes,1,opt,name=source" json:"source,omitempty"`
	Metric           *string      `protobuf:"bytes,2,opt,name=metric" json:"metric,omitempty"`
	Timestamp        *int64       `protobuf:"varint,3,opt,name=timestamp" json:"timestamp,omitempty"`
	Value            *Datum       `protobuf:"bytes,4,opt,name=value" json:"value,omitempty"`
	MetricType       *MetricType  `protobuf:"varint,5,opt,name=metricType,enum=com.signalfx.metrics.protobuf.MetricType" json:"metricType,omitempty"`
	Dimensions       []*Dimension `protobuf:"bytes,6,rep,name=dimensions" json:"dimensions,omitempty"`
	XXX_unrecognized []byte       `json:"-"`
}
```
I assume that `Properties` that you mention comes from the client lib, let me see what it is translated to.

File: receiver/signalfxreceiver/receiver.go (hunk 1 above)
Comment (tigrannajaryan) [reply receiver/signalfxreceiver/receiver.go]: Sure, up to you on the metric, I don't have a strong opinion on this.

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go (hunk 3 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: Yes, the zero value - will add explicitly.

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go (hunk 2 above)
Comment (tigrannajaryan) [reply receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: OK, ignore this.

Looks like it was removed from Protobuf definition https://github.com/signalfx/com_signalfx_metrics_protobuf/commit/5e41ba1725d48aba7aa292d3ee4d0c0e6ddd6937

I just had an old version of SmartAgent source code which was refering to the version of the Protobuf which had this field.

However, this makes it unclear to me how are resource attributes reported? `Dimensions` are supposedly the identity of the metric, i.e. the label keys. Where do resource attributes like hostname go? Also in `Dimensions`?

Comment (pjanotti) [reply receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: I found `Properties` on the event type:
```golang
type Event struct {
	EventType        *string        `protobuf:"bytes,1,req,name=eventType" json:"eventType,omitempty"`
	Dimensions       []*Dimension   `protobuf:"bytes,2,rep,name=dimensions" json:"dimensions,omitempty"`
	Properties       []*Property    `protobuf:"bytes,3,rep,name=properties" json:"properties,omitempty"`
	Category         *EventCategory `protobuf:"varint,4,opt,name=category,enum=com.signalfx.metrics.protobuf.EventCategory" json:"category,omitempty"`
	Timestamp        *int64         `protobuf:"varint,5,opt,name=timestamp" json:"timestamp,omitempty"`
	XXX_unrecognized []byte         `json:"-"`
}
```

File: receiver/signalfxreceiver/receiver.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply receiver/signalfxreceiver/receiver.go]: I see. So this refers to the fact the consumer is provided in `New` and `StartMetricsReception` is going to send to that consumer.

Perhaps rephrase it to make it clearer, I was a bit confused.

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go (hunk 2 above)
Comment (tigrannajaryan) [reply receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: Perhaps check with SFx folks.

File: receiver/signalfxreceiver/receiver.go (hunk 1 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: Going with a custom metric now, will update when we have something uniform on core.

File: receiver/signalfxreceiver/receiver.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: Either way it doesn't matter much, will remove to not raise Qs by people reading the code (I think this is the second time).

File: receiver/signalfxreceiver/receiver.go (hunk 1 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: Taking that back, I will add the metric late and will keep as Debug level for now.

**Review by tigrannajaryan — APPROVED** [review]

File: receiver/signalfxreceiver/receiver.go [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+	server       *http.Server
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+var _ receiver.MetricsReceiver = (*sfxReceiver)(nil)
+
+func (r *sfxReceiver) MetricsSource() string {
+	const metricsSource string = "SignalFx"
```
Comment (owais) [inline receiver/signalfxreceiver/receiver.go]: Is there a reason for declaring this inside the func?

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+	case sfxpb.MetricType_GAUGE:
+		// Numerical: Periodic, instantaneous measurement of some state.
+		descType = metricspb.MetricDescriptor_GAUGE_DOUBLE
+		if sfxDatum.IntValue != nil {
+			descType = metricspb.MetricDescriptor_GAUGE_INT64
+		}
+
+	case sfxpb.MetricType_COUNTER:
+		// Numerical: Count of occurrences. Generally non-negative integers.
+		fallthrough
```
Comment (owais) [inline receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: Are we counting this cumulative counter as well? Would `case COUNTER, COMULATIVE_COUNTER` be better/more-obvious than fallthrough?

File: receiver/signalfxreceiver/receiver.go (hunk 7 above)
Comment (pjanotti) [reply receiver/signalfxreceiver/receiver.go]: Actually no :) will remove

File: receiver/signalfxreceiver/signalfxv2_to_metricdata.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+func convertTimestamp(msec int64) *timestamp.Timestamp {
+	if msec == 0 {
+		return nil
+	}
+
+	ts := &timestamp.Timestamp{
+		Seconds: msec / 1e3,
+		Nanos:   int32(msec%1e3) * 1e6,
+	}
```
Comment (owais) [inline receiver/signalfxreceiver/signalfxv2_to_metricdata.go]: Seeing this again and again. We should add it as a time util func somewhere after this PR is merged.

**Review by owais — APPROVED** [review]
